File: Languages/python/algorithms-dsa/leetcode/linked-list/merge_two_sorted_lists.py
"""
LeetCode #21 - Merge Two Sorted Lists
Topic: Linked List
Difficulty: Easy

Merge two sorted linked lists iteratively.

Time Complexity: O(n + m)
Space Complexity: O(1)
"""


class ListNode:
    def __init__(self, val: int = 0, next: "ListNode | None" = None):
        self.val = val
        self.next = next


def merge_two_lists(list1: ListNode | None, list2: ListNode | None) -> ListNode | None:
    dummy = ListNode(0)
    curr = dummy
    while list1 and list2:
        if list1.val <= list2.val:
            curr.next = list1
            list1 = list1.next
        else:
            curr.next = list2
            list2 = list2.next
        curr = curr.next
    curr.next = list1 or list2
    return dummy.next


def to_list(head: ListNode | None) -> list[int]:
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


def from_list(vals: list[int]) -> ListNode | None:
    dummy = ListNode(0)
    curr = dummy
    for v in vals:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


if __name__ == "__main__":
    assert to_list(merge_two_lists(from_list([1, 2, 4]), from_list([1, 3, 4]))) == [1, 1, 2, 3, 4, 4]
    assert to_list(merge_two_lists(from_list([]), from_list([]))) == []
    assert to_list(merge_two_lists(from_list([]), from_list([0]))) == [0]
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/linked-list/merge_two_sorted_lists.cpp
/*
 * LeetCode 21 - Merge Two Sorted Lists
 * Topic: Linked List
 * Difficulty: Easy
 *
 * Iterative merge using a dummy head node.
 * Time: O(n + m)
 * Space: O(1)
 */
#include <iostream>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    while (l1 && l2) {
        if (l1->val <= l2->val) { curr->next = l1; l1 = l1->next; }
        else { curr->next = l2; l2 = l2->next; }
        curr = curr->next;
    }
    curr->next = l1 ? l1 : l2;
    return dummy.next;
}

ListNode* buildList(initializer_list<int> vals) {
    ListNode dummy(0);
    ListNode* curr = &dummy;
    for (int v : vals) { curr->next = new ListNode(v); curr = curr->next; }
    return dummy.next;
}

bool equals(ListNode* a, initializer_list<int> vals) {
    for (int v : vals) { if (!a || a->val != v) return false; a = a->next; }
    return a == nullptr;
}

int main() {
    assert(equals(mergeTwoLists(buildList({1,2,4}), buildList({1,3,4})), {1,1,2,3,4,4}));
    assert(equals(mergeTwoLists(nullptr, nullptr), {}));
    assert(equals(mergeTwoLists(nullptr, buildList({0})), {0}));

    cout << "All tests passed!" << endl;
    return 0;
}
